### DISCUSSION TITLE
Azure Pipelines Scaler - Placeholder Agent Config

### DISCUSSION THREAD

**Discussion by thepaulmacca** [kedacore/keda#2649] [Q&A / Need Help]
1) I'm attempting to scale to zero and set up a placeholder agent as per the [docs](https://keda.sh/blog/2021-05-27-azure-pipelines-scaler/#placeholder-agent), but I could do with some help on the agent `start.sh` script

```bash
#!/bin/bash
set -e

if [ -z "$AZP_URL" ]; then
  echo 1>&2 "error: missing AZP_URL environment variable"
  exit 1
fi

if [ -z "$AZP_TOKEN_FILE" ]; then
  if [ -z "$AZP_TOKEN" ]; then
    echo 1>&2 "error: missing AZP_TOKEN environment variable"
    exit 1
  fi

  AZP_TOKEN_FILE=/azp/.token
  echo -n $AZP_TOKEN > "$AZP_TOKEN_FILE"
fi

unset AZP_TOKEN

if [ -n "$AZP_WORK" ]; then
  mkdir -p "$AZP_WORK"
fi

rm -rf /azp/agent
mkdir /azp/agent
cd /azp/agent

export AGENT_ALLOW_RUNASROOT="1"

cleanup() {
  if [ -e config.sh ]; then
    print_header "Cleanup. Removing Azure Pipelines agent..."

    ./config.sh remove --unattended \
      --auth PAT \
      --token $(cat "$AZP_TOKEN_FILE")
  fi
}

print_header() {
  lightcyan='\033[1;36m'
  nocolor='\033[0m'
  echo -e "${lightcyan}$1${nocolor}"
}

# Let the agent ignore the token env variables
export VSO_AGENT_IGNORE=AZP_TOKEN,AZP_TOKEN_FILE

print_header "1. Determining matching Azure Pipelines agent..."

AZP_AGENT_RESPONSE=$(curl -LsS \
  -u user:$(cat "$AZP_TOKEN_FILE") \
  -H 'Accept:application/json;api-version=3.0-preview' \
  "$AZP_URL/_apis/distributedtask/packages/agent?platform=linux-x64")

if echo "$AZP_AGENT_RESPONSE" | jq . >/dev/null 2>&1; then
  AZP_AGENTPACKAGE_URL=$(echo "$AZP_AGENT_RESPONSE" \
    | jq -r '.value | map([.version.major,.version.minor,.version.patch,.downloadUrl]) | sort | .[length-1] | .[3]')
fi

if [ -z "$AZP_AGENTPACKAGE_URL" -o "$AZP_AGENTPACKAGE_URL" == "null" ]; then
  echo 1>&2 "error: could not determine a matching Azure Pipelines agent - check that account '$AZP_URL' is correct and the token is valid for that account"
  exit 1
fi

print_header "2. Downloading and installing Azure Pipelines agent..."

curl -LsS $AZP_AGENTPACKAGE_URL | tar -xz & wait $!

source ./env.sh

print_header "3. Configuring Azure Pipelines agent..."

./config.sh --unattended \
  --agent "${AZP_AGENT_NAME:-$(hostname)}" \
  --url "$AZP_URL" \
  --auth PAT \
  --token $(cat "$AZP_TOKEN_FILE") \
  --pool "${AZP_POOL:-Default}" \
  --work "${AZP_WORK:-_work}" \
  --replace \
  --acceptTeeEula & wait $!

print_header "4. Running Azure Pipelines agent..."

trap 'cleanup; exit 130' INT
trap 'cleanup; exit 143' TERM

# To be aware of TERM and INT signals call run.sh
# Running it with the --once flag at the end will shut down the agent after the build is executed
./run.sh --once & wait $!
```

Can someone tell me what I need to remove from here please, to make the scaling to zero work? I'm quite new to this, so any help would be appreciated

2) Another question I have is, how long should it take before an agent scales down?

Thanks

**Comment by Frettarix** [comment]
Heya,

So for the first question what I understand is: You're using KEDA autoscaler, and you want to have a placeholder there but the Microsoft original script does not allow this, correct?

I'd first recommend to update your code and synchronize it with what MS suggests: https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops
(If you want to run --once, pass this as an argument to your container, you do not need to edit the start.sh for this purpose ;) )

And then to make the agent NOT deregister itself after removal you can simply edit the cleanup() part of the start.sh:
```
cleanup() { 
  if [ -e config.sh ]; then 
    if [ -z "$AZP_PLACEHOLDER" ]; then 
      print_header "Cleanup. Removing Azure Pipelines agent..." 

      # If the agent has some running jobs, the configuration removal process will fail. 
      # So, give it some time to finish the job. 
      while true; do 
        ./config.sh remove --unattended --auth PAT --token $(cat "$AZP_TOKEN_FILE") && break 

        echo "Retrying in 30 seconds..." 
        sleep 30 
      done 
    else 
      print_header "Cleanup skipped as Agent is marked as a placeholder... this option should be removed if Azure DevOps allows queueing to empty Agent Pools." 
    fi 
  fi 
} 
```
Now when you run this container and you pass ANY value to "AZP_PLACEHOLDER" it will skip the cleanup task and you have an offline agent. In our environment we also change its AZP_AGENT_NAME to "KEDA-Placeholder" to make sure we know why it's there.

For the second question I do not have an answer, we use ScaledJob, so we get a job per item in the Agent Pool, when the job is done, the agent deregisters. But for things to be allowed into this queue we do need the placeholder agent.

Hopefully this will help you on your way ;)

/Fret

**Reply by nikhil-rajan-neu** [reply]
@Frettarix .. i am also trying to the same thing . I understand that we need to modify the clean up script . but how do you make an offline agent . How can we do that

**Reply by Frettarix** [reply]
Hi @nikhil-rajan-neu , to get an offline agent (or placeholder agent) simply deploy an agent where the clean-up script has been disabled. When the cleanup script is not disabled > the agent will be removed from the agent pool when deleted.
When the cleanup script is disabled > the agent will not be removed from the agent pool when deleted.

Please see the provided changes to the start.sh.

So in the end to deploy one KEDA agent:
- Deploy an agent without cleanup and delete (this will get you an offline agent)
- Deploy the KEDA agent you actually want

Hopefully that'll work for you ;)

**Reply by nikhil-rajan-neu** [reply]
@Frettarix .. ok i get in now .. 
Another related question .. we are having windows agents too ..hosted on windows node .. 
do you happen to know the equivalent to the run once in the windows start.ps1 file .. couldn't find anything in the documentation .. its only available for Linux based agents

**Reply by Frettarix** [reply]
@nikhil-rajan-neu  Sorry mate, haven't worked with Windows agents before. Perhaps you can look at what I did with the shell script and see if a similar implementation is possible for the PWSH script ;)

**Reply by Jeet-14** [reply]
Were you able to bypass default parallelism? using scaledJobs?
As per the [blog](https://keda.sh/blog/2021-05-27-azure-pipelines-scaler/)
KEDA will autoscale to the maximum defined in the ScaledObject and does not limit itself to the parallel jobs count defined for the Azure DevOps organization.

I am getting following error message in ADO pipeline: while trying to run multiple pipelines at the same time on pool.

Err msg
```bash
This agent request is not running because you have reached the maximum number of requests that can run for parallelism type 'Self-Hosted Private'. Current position in queue: 1. Current parallelism: 1. Max Parallelism for parallelism type 'Self-Hosted Private': 1
```

**Reply by Frettarix** [reply]
You can overwrite the default parallism with parameters. Change it in your scaledJob ;)

**Reply by Jeet-14** [reply]
Ok here is the suitation: I;ve set     `parallelism: 5` in scaledJob manifest and I can see five agents in my pool.
![image](https://github.com/kedacore/keda/assets/24683251/b4ec3a02-4659-4d93-aaf8-194baeb880d8)

Inside pipeline, am not running anything fancy, just hello world and sleep for 300 sec.
![image](https://github.com/kedacore/keda/assets/24683251/57f744e4-e0af-41a0-b5d0-9f3a0adf1a12)
I was expecting them to run in parallel, per pod one job but those are in queue waiting for first to be complete.

---
Question is:  will setting-up 'parallelism' in scaledJob can bypass restriction of azure org level 'Free parallel jobs'?
What I want is to run pipeline jobs in parallel but not the paid one.
![image](https://github.com/kedacore/keda/assets/24683251/d0f26b19-c625-4087-906c-c5c7037249ac)

Let me know if i misinterpret the blog!
